### DISCUSSION TITLE
it is not working for JSON/API Data, but MathJax is working. It just works for objects only

### DISCUSSION THREAD

**Discussion by dibasdauliya** [KaTeX/KaTeX#2782] [Q&A] [answered]
**Describe the bug:**
Because it is same file, it is working for objects but for JSON/API it is not working. I replace the code with MathJax and it is working. I have included all the example below:

**KaTex (with JS Object it is working) but (it is not working for JSON)**
**This is the example for JS object (it is working)**
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css" integrity="sha384-AfEj0r4/OFrOo5t7NnNe46zW/tFgW6x/bCJG8FqQCEo3+Aro6EYUG4+cU+KJWu/X" crossorigin="anonymous"/></link>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/contrib/auto-render.min.js"></script>
  </head>
  <body>
    $\frac y3$ $\frac y3$
    <div id="math"></div>

    <main id="main"></main>
  </body>
  <script type="text/javascript" src="main.js"></script>
  <script>
          renderMathInElement(document.body, {
              delimiters: [
                  {left: "$$", right: "$$", display: true},
                  {left: "$", right: "$", display: false},
                  {left: "\\(", right: "\\)", display: false},
                  {left: "\\[", right: "\\]", display: true},
                  {left: "\\begin{equation}", right: "\\end{equation}", display: true}
              ]
          });
  </script>
</html>
```
```js
const main = document.getElementById('main')

const data = {
  angleBtnS: [
    {
      question: 'Prove that .. $\\frac x2$ + $\\frac y3$',
      date: 'SEE 2073',
      ans: '11'
    },
    {
      question:
        'If the lines 3x + my = 5 and $\\frac x2$ + $\\frac y3$ = 1 are parallel to each other, find the value of m.',
      date: '22',
      ans: '<h1>hihi</h1>'
    },
    {
      question: '$\\frac x2$ + $\\frac y3$ - $\\frac x2$ + $\\frac y3$',
      date: 'date',
      ans: 'ans ans ans'
    }
  ]
}

function setData() {
  main.innerHTML = ''

  data.angleBtnS.forEach((dta, index) => {
    const dataEl = document.createElement('div')
    dataEl.classList.add('card')

    dataEl.innerHTML = `
      <h2>${index + 1 + ') '}${dta.question}</h2><span>${dta.date}</span>
      <div class="ans">${dta.ans}</div>
      `
    main.appendChild(dataEl)
  })
}
setData()
```

**this is the example of JSON/From API (it is not working)**
**all the code are same but I replace the config with MathJax and for Mathjax same is working fine.**
**this is the example of KaTeX, next one is of MathJax**
```html
html file is exactly same as above
```
this is js file: 

```js
async function fetchData() {
  const res = await fetch('main.json') //main.json file is same as of the js object 'const data = ...' which is shown in above example
  const resData = await res.json()
  setData(resData.angleBtnStraight)
}

function setData(data) {
  main.innerHTML = ''

  data.forEach((dta, index) => {
    const dataEl = document.createElement('div')
    dataEl.classList.add('card')

    dataEl.innerHTML = `
      <h2>${index + 1 + ') '}${
      dta.question
    }</h2><span>${dta.date}</span>
      <div class="ans">${dta.ans}</div>
      `
    main.appendChild(dataEl)
  })
}
fetchData()
```

**FINALLY, this the mathjax code which is working for JSON:**
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
      MathJax.Hub.Config({
          tex2jax: {inlineMath: [['$','$']]},
          "HTML-CSS": { linebreaks: { automatic: true }, mtextFontInherit: true, matchFontHeight: false, mathmlSpacing: true },
          CommonHTML: { linebreaks: { automatic: true }, mtextFontInherit: true, matchFontHeight: false, mathmlSpacing: true },
          SVG: { linebreaks: { automatic: true }, mtextFontInherit: true }
        });
    </script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_HTML-full"></script>
  </head>
  <body>
    $\frac y3$ $\frac y3$
    <div id="math"></div>

    <main id="main"></main>
  </body>
  <script type="text/javascript" src="main.js"></script>
</html>
```
js file:
`
js file is same as above (just above this one i.e. 2nd example)
`

**Comment by edemaine** [comment, marked as answer]
I would recommend putting the call to KaTeX (`renderMathInElement`) at the end of your JavaScript file `main.js`.  Then it seems to work well.  See [this CodePen demo](https://codepen.io/edemaine/pen/NWbrjRY).

On the other hand, the code you show *should* work &mdash; I believe `<script>` tags are supposed to run in order, unless they are marked `defer` or `async`.  I tried running your code directly as listed above, and I got some rendering defects (fraction lines invisible, not sure why) but it did render all the math.  So it's difficult to tell what's going wrong in your setup.

**Reply by dibasdauliya** [reply]
It is working in objects but not working for JSON. please see [this demo in codePen](https://codepen.io/dibasdauliya/pen/oNYYzJm) it is not working :(
I also put the call to KaTeX (`renderMathInElement`) at the end of JavaScript file main.js

**Reply by edemaine** [reply]
Ah, now I see the issue. You have two uses of `await` in `fetchData`, and KaTeX is rendering while those promises happens.  A simple approach would be to move the call to KaTeX's renderer to the end of `setData`, so that it's called after everything gets fetched and put in the DOM.

**Reply by dibasdauliya** [reply]
Oh Yeah! Thank you so much @edemaine 
Anyways, is this good method? any recommendations for me?

**Reply by edemaine** [reply]
Yes, this seems reasonable. (I would put the KaTeX rendering in its own function.) Alternatively, you could prerender the math and store the rendered HTML in the JSON file; then you don't have to do anything client-side.

**Reply by dibasdauliya** [reply]
You mean I should keep KaTeX rendering in function (lets suppose `function callKatex() {...}`) and call inside `setData` as `callKatex()`?
2nd recommendation, you mean instead of `question: 'If the lines 3x + my = 5 and $\\frac x2$ + $\\frac y3$ = 1 are parallel to each other, find the value of m.',` I should write `question: 'If the lines 3x + my = 5 and <span ..></..> + <span ..></..> = 1 are parallel to each other, find the value of m.',` by inspecting in dev tools? can you please explain more (shy emoji)

**Reply by edemaine** [reply]
Yes to the first point.

No, definitely don't use dev tools.  I meant that you could write a NodeJS program that translates your JSON file using LaTeX notation into another JSON file that uses HTML notation. But this is just an optimization, and not important.

**Reply by dibasdauliya** [reply]
oh. can you please suggest the name of npm package that you like? I want to translate JSON file using LaTeX notation into another JSON file. 

It would be very awesome if you could provide code to translate JSON 😅 I searched in the internet. got to know about `tex4ht`, `itexToMML`, searched about its implementation but don't know how to start :(
It'd be so great if you help in this.

**Reply by edemaine** [reply]
There isn't a package for this. The idea would be to do it via `katex.renderToString`. But again, this is not necessary, just an optimzation.

**Reply by dibasdauliya** [reply]
Nono I want to optimize it, learn it. 😅 can you please help with this? `katex.renderToString`? 🤔

I read [this](https://katex.org/docs/api.html#server-side-rendering-or-rendering-to-a-string) but not getting good idea (to convert json)

**Reply by dibasdauliya** [reply]
Should I render the each required math in console by using `katex.renderToString` and copy the html in json? (if I want to optimize). Is there any recommendation?

I also know some reactJS, as you said before that it can be done form nodeJS, can you help me to accomplish it? It'd make site load faster and also I could get new thing to learn

**Reply by edemaine** [reply]
Sorry, but this is beyond the scope of what I have time to help with. Best of luck!
